[2단계 - 장기] 체체(김진영) 미션 제출합니다.

CheChe903 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요, 감기에 걸려서 컨디션 관리하느라 시간이 좀 늦었습니다! 리뷰 잘부탁드립니다.
또한 아직 데이터베이스 관리 (Dao)에 대해 개념이 잡혀있지 않아 먼저 PR날리고 고치자는 생각으로 코드가 많이 더럽습니다 (또한 dao 테스트도 적지 않은 부분이 많습니다.) . 많이 도와주세요 !



>DB 의 저장, 불러오기가 쉽다 는 부수적인 효과라고 생각해도 될까요?
객체의 설계에 있어서 DB 에 관한 부분을 고려한다면 DB 에 의존적인 설계가 될 것 같아요
물론 프로그램은 보통 저장소(DB)에 데이터를 저장하여 사용하기에 아예 고려를 안할 수는 없지만
객체 설계 >>>> DB 고려 의 중요도라고 생각해요.
체체는 어떻게 생각하시나요?

부수적인 효과라고 전달드리고 싶었습니다! 저도 찰리와 같은 의견입니다.

>Optional 을 활용해주셨군요
아래 글의 3번 항목을 읽고 체체의 의견을 공유해주면 좋겠어요~

그렇군요! Optional에 대해 제 기준이 없었는데 이떄문에 기준이 잡혔습니다!

>Chariot 과 공통된 부분으로 보이네요 :)
추상화 계층을 만들어서 중복된 부분을 추출해보는것은 어떻게 생각하시나요?

 고민하느라 굉장히 많은 시간이 들었어요. 
 결론적으로 기물들이 경로를 생성하는 방식이나 그 경로가 올바른지는 같을 수 있지만, 각각 다른 책임이라고 생각합니다! 
 물론 찰리가 집어준 부분에서 차와 포의 로직이 같을 순 있지만, 엄연히 차의 생성방식, 포의 생성방식이라고 기준을 두었습니다!
 이 때문에 추상 클래스를 사용하는 의미를 좀 잡은 것 같아요. 예시로, 기물이 움직이는 것이나, 같은 팀인가에 대한 것은 모두 바뀌어야 하기 때문에 추출하여 추상 클래스에 구현하였지만, 각 기물의 움직임 전략은 모두 다르게 바뀔 수 있기 때문에 추상 메서드로 두었습니다!

## 질문사항

1. test
Dao, 즉 데이터베이스와 연결되는 부분들을 테스트해야할까요 ? 
만약 한다면 어떻게 두 사이를 구분할 수 있을까요? 시도해보았지만, id부분에서 저장하는데 어려움을 겪었습니다.

(Auto Increment를 사용했지만, 테이블을 삭제하지 않는 한 해당 부분은 고정되어 새로 생성해도 그 이후 int 값을 가집니다. 이 때문에 테스트가 깨지는 경우도 존재합니다. 테스트를 작성하지 않은 부분은 이 문제때문에 찰리에게 의견을 듣고 고치고 싶습니다.)

제가 생각한 방법은

FakeDao 생성 후 Map이나 다른 자료구조 이용 
하지만 생성하는데 많은 시간이 소모됩니다. 또한 데이터베이스에 저장되는 테스트인데, 자료구조로 퉁친다면 뭔가 이상합니다.

FakeDao 생성 후 testDB 생성
위와 비슷하지만 테스트용 데이터베이스를 생성하여 테스트합니다. auto Increment 문제를  id까지 쿼리문에 넣는 방식으로 해결합니다. 하지만 여전히 생성 비용과 관리 비용이 많은 것 같습니다.


2. 피스를 옮길 때마다 저장? 한 번에 저장?
게임을 진행하면서 해당 사항을 저장하면서 가야할까요, 아니면 3번 기능(게임 종료)을 이용했을 때 일괄적으로 저장해야할까요 ?

3. 추상 클래스의 한계
데이터베이스에 저장하기 위해서는 해당 기물의 종류를 알아야했습니다. 그렇다면 추상 클래스를 사용하면서 데이터베이스에 저장하기 위해서는 getClass로 변환해주는 작업이 필수인 걸까요 ?



<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - 향상된 장기 게임 플레이: 특정 기물(졸, 사, 왕, 차, 포)의 이동 규칙이 업데이트되어 대각선 이동 등 새로운 움직임이 적용되었습니다.
  - 게임 상태 및 점수 체계가 보강되어, 게임 결과와 팀별 점수가 명확하게 표시됩니다.
  - 사용자 인터랙션이 개선되어 게임 중 중지 및 신규 게임 시작 옵션이 추가되었습니다.
  - 데이터베이스 사용법 및 SQL 명령어가 README에 추가되었습니다.
  - MySQL 데이터베이스와 관련된 새로운 서비스 및 테이블이 도입되었습니다.

- **Documentation**
  - README에 게임 결과 조건 및 데이터베이스 설정 방법 등 상세한 사용자 안내가 추가되었습니다.

- **Chores**
  - MySQL과 phpMyAdmin 구동을 위한 Docker Compose 설정 파일이 도입되었습니다.

- **Tests**
  - 게임 로직, 데이터베이스 연결, 저장소 기능을 검증하는 포괄적인 테스트들이 추가되었습니다.
  - 새로운 테스트 클래스가 추가되어 게임 상태 및 턴 관리 기능을 검증합니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

CheChe903 committed:
docs(README): 2단계 요구사항 업데이트

CheChe903 committed:
refactor(Cannon): 메서드 분리

CheChe903 committed:
test: 기물 경로 및 이동에 따른 테스트 추가

CheChe903 committed:
refactor(Piece): 공통 부분 메서드 Piece로 공통화

CheChe903 committed:
docs(README): 요구 사항 업데이트

CheChe903 committed:
docs(README): 요구 사항 업데이트

CheChe903 committed:
feat(Piece): 장기 점수 및 게임 종료 , 단위 테스트 구현

CheChe903 committed:
feat(Controller): 게임 종료 기능 구체화

CheChe903 committed:
docs(README): 요구사항 업데이트

CheChe903 committed:
refactor(Piece): 추상화 수정 및 General, Guard 궁성 움직임 추가

CheChe903 committed:
feat(Cannon, Chariot): 궁성 대각선 움직임 및 단위 테스트 구현

CheChe903 committed:
feat(Position): 특정 점 대각선 이동 및 단위 테스트 구현

CheChe903 committed:
refactor(Piece): 메서드 리팩터링

CheChe903 committed:
docs(README): 요구사항 업데이트

CheChe903 committed:
docs(Docker): Docker 및 데이터베이스 설정

- db 폴더는 ignore 처리합니다.

CheChe903 committed:
refactor(Board): 객체명 변경

- Board -> Game 으로 변경합니다.

CheChe903 committed:
feat(Dao): 데이터베이스 로직 추가

CheChe903 committed:
feat(Dao): 데이터베이스 로직 추가

CheChe903 committed:
docs(README): 데이터베이스 사용법 업데이트

CheChe903 committed:
docs(README): 출력 예시 수정

Gomding commented on src/main/java/janggi/domain/piece/Chariot.java:
```diff
@@ -11,6 +20,19 @@
 
 public class Chariot extends Piece {
```
> 고민하느라 굉장히 많은 시간이 들었어요.
> 결론적으로 기물들이 경로를 생성하는 방식이나 그 경로가 올바른지는 같을 수 있지만, 각각 다른 책임이라고 생각합니다!
> 물론 찰리가 집어준 부분에서 차와 포의 로직이 같을 순 있지만, 엄연히 차의 생성방식, 포의 생성방식이라고 기준을 두었습니다!
> 이 때문에 추상 클래스를 사용하는 의미를 좀 잡은 것 같아요. 예시로, 기물이 움직이는 것이나, 같은 팀인가에 대한 것은 모두 바뀌어야 > 하기 때문에 추출하여 추상 클래스에 구현하였지만, 각 기물의 움직임 전략은 모두 다르게 바뀔 수 있기 때문에 추상 메서드로 두었습니다!

추상 클래스로 계층을 두는것도 고려해볼만 해서 드린 코멘트였어요!
Piece > 직선이동Piece(?) > Chariot, Cannon
하지만 약간 복잡해진다는 단점도 있을것 같네요 :)

> 엄연히 차의 생성방식, 포의 생성방식이라고 기준을 두었습니다! 

해당 답변도 좋습니다!! 👍 👍

CheChe903 replied:
해당 코멘트 덕에 저만의 기준을 잡을 수 있었습니다 :) 감사합니다.

Gomding commented on src/test/java/janggi/dao/DatabaseConnectionTest.java:
```diff
@@ -0,0 +1,25 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class DatabaseConnectionTest {
```
데이터베이스 커넥션까지 꼼꼼하게 테스트하셨군요 😃 
어떤것때문에 커넥션 맺는 부분도 테스트해야겠다고 느끼셨는지 궁금해요!

CheChe903 replied:
만약 해당 테스트가 없고, dao 테스트가 모두 실패한다면 커넥션 테스트가 문제인 지 한 번에 파악하기 어렵다고 판단되어 테스트 하였습니다!

Gomding commented on src/main/java/janggi/dao/DatabaseConnection.java:
```diff
@@ -0,0 +1,49 @@
+        return instance;
+    }
+
+    public Connection getConnection() {
+        try {
+            return DriverManager.getConnection(URL, USERNAME, PASSWORD);
+        } catch (final SQLException e) {
+            System.err.println("DB 연결 오류:" + e.getMessage());
+            e.printStackTrace();
+            return null;
+        }
+    }
```
[해당 코멘트](https://github.com/woowacourse/java-janggi/pull/100/files#r2019779171)는 꼭 고민해보시면 좋겠어요 :)
연결 실패 시 에러 출력만하고 null을 반환하는것은 어떤 문제가 생길 수 있을까요?

CheChe903 replied:
@coderabbitai
그렇다면 find query가 존재하지 않으면 어떻게 처리해야할까?

Gomding commented on src/main/java/janggi/dao/DatabaseConnection.java:
```diff
@@ -0,0 +1,49 @@
+    private static final String OPTION = "?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=UTC";
+    private static final String USERNAME = "root";
+    private static final String PASSWORD = "root";
+    private static final String URL = "jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION;
```
해당 코멘트는 현재 단계에서 신경쓰지 않으셔도 좋아요!

Gomding commented on src/main/java/janggi/dao/DatabaseConnection.java:
```diff
@@ -0,0 +1,49 @@
+            }
+        }
+    }
+}
```
[요것도](https://github.com/woowacourse/java-janggi/pull/100/files#r2019779174) 현재 단계에서는 적용하지 않으셔도 좋습니다 ㅎㅎ
단 HikariCP에 대해 학습정도는 해봐도 좋아요~! 
-> 현재 단계에서는 깊게 이해하지 못해도 괜찮아요! 그냥 이런게 있구나 정도만 알고 계서도 좋습니다 😄

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -61,15 +120,28 @@ private boolean canMove(final Set<Route> possibleRoutes, final Position destinat
     }
+
+    private boolean stopGame() {
+        if (inputView.inputStopGame() == Y & inputView.inputStopGame() == Y) {
```
요건 입력을 두 번 받는 이유가 궁금해요!

> 서로가 게임 종료에 동의하면 남아있는 기물의 점수를 계산 후 출력한다.

해당 요구사항을 구현하신걸까요?
맞다면 `서로` 라는것이 표현되면 좋겠어요 :)

CheChe903 replied:
메서드명만 수정해보았는데 괜찮을까요 ?
현재로서는 콘솔 애플리케이션이라 이렇게 할 수 밖에 없는 것 같아요.

[fdab2a7](https://github.com/woowacourse/java-janggi/pull/100/commits/fdab2a777519c871f10509bf6bf155a930e645b1)

Gomding replied:
음 사용자 입장에서는 단순 두 번 입력하는것처럼 나와서 드린 코멘트였어요! 
메서드 이름과 입력에 대한 안내 방식을 고민해봐도 좋을 것 같네요 :)

![image](https://github.com/user-attachments/assets/bc766b6e-8c1e-4170-98ef-fe53aea0b38b)

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -18,39 +28,88 @@ public class JanggiController {
+    }
+
+    private Game generateBoard() {
+        if (pieceDao.findAllPieces().isEmpty()) {
+            pieceDao.deleteAllPieces();
+            turnDao.deleteTurn();
+            return generateNewBoard();
+        }
+        if (inputView.inputNewGame() == Y) {
+            pieceDao.deleteAllPieces();
+            turnDao.deleteTurn();
+            return generateNewBoard();
+        }
+        return new Game(new Pieces(pieceDao.findAllPieces()), turnDao.findTurn());
```
> pieceDao.findAllPieces()

데이터 베이스 조회하는 메서드를 두 번 호출하고 있는데 횟수를 줄여보면 좋을 것 같아요!
데이터베이스 접근은 비용이 비싼 행위입니다 :)
왜 비싼 행위인지 학습해보시면 좋겠어요!

CheChe903 replied:
찾아보니 많은 비용이 든다고 하네요! (네트워크, 트랜잭션이 필요한 경우 트랜잭션 비용 등) 인지하도록 하겠습니다 !

[ff0cb88](https://github.com/woowacourse/java-janggi/pull/100/commits/ff0cb884b41cf20858401a91e29de0d423323335)

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -18,39 +28,88 @@ public class JanggiController {
+    }
+
+    private Game generateBoard() {
+        if (pieceDao.findAllPieces().isEmpty()) {
```
단순 존재하는지 확인하는 거라면 exist 나 count 쿼리를 활용해보시는건 어떻게 생각하세요? 😃

CheChe903 replied:
해당 부분은 위 코멘트에 의해 고쳐지면서 존재 확인 뿐만아니라 입력에 따라 사용되니 그대로 두었습니다!
다만, 조회하는 부분에서는 count를 이용할 수 있네요 :)

Gomding commented on src/main/java/janggi/dao/PieceDao.java:
```diff
@@ -0,0 +1,138 @@
+
+    public void addPiece(final Piece piece) {
+        final var query = "INSERT INTO piece (piece_type, team_id, position_id) VALUES (?, ?, ?)";
+        try (final var connection = databaseConnection.getConnection();
```
`try-with-resource` 사용해주셨군요!! 👍 👍 👍

Gomding commented on src/test/java/janggi/dao/PieceDaoTest.java:
```diff
@@ -0,0 +1,62 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class PieceDaoTest {
```
> ## Dao, 즉 데이터베이스와 연결되는 부분들을 테스트해야할까요 ?

체체는 테스트가 필요하다고 생각하셔서 작성을 시도하신거겠죠? 😃 
필요하다고 느낀 이유는 무엇이었을까요?

> 만약 한다면 어떻게 두 사이를 구분할 수 있을까요? 시도해보았지만, id부분에서 저장하는데 어려움을 겪었습니다.
> (Auto Increment를 사용했지만, 테이블을 삭제하지 않는 한 해당 부분은 고정되어 새로 생성해도 그 이후 int 값을 가집니다. 이 때문에 테스트가 깨지는 경우도 존재합니다. 테스트를 작성하지 않은 부분은 이 문제때문에 찰리에게 의견을 듣고 고치고 싶습니다.)

테스트용 DB 를 운영하는 방법
또는 id에 의존하지 않고 테스트하는 방법이 있을 것 같은데
어떤 어려움이 있는지 예시도 보여주시면 좋을 것 같아요!

> 제가 생각한 방법은
> 
> FakeDao 생성 후 Map이나 다른 자료구조 이용
> 하지만 생성하는데 많은 시간이 소모됩니다. 또한 데이터베이스에 저장되는 테스트인데, 자료구조로 퉁친다면 뭔가 이상합니다.
> 
> FakeDao 생성 후 testDB 생성
> 위와 비슷하지만 테스트용 데이터베이스를 생성하여 테스트합니다. auto Increment 문제를 id까지 쿼리문에 넣는 방식으로 해결합니다. 하지만 여전히 생성 비용과 관리 비용이 많은 것 같습니다.

FakeDao 를 사용하여 테스트한다는것은 어떤 의미인지 이해못했어요 🤔 
PieceDao 를 테스트할 건데 FakePieceDao 를 구현해서 테스트한다는것일까요?

CheChe903 replied:
>필요하다고 느낀 이유는 무엇이었을까요?

먼저, 실제 서버를 이용하며 사용자의 트래픽을 받는다고 가정합니다.
이때 다른 팀원이 Dao 테스트 없이 배포를 한 후 데이터들이 저장이 안되는 문제가 발생할 수 있습니다! 
이러한 이유로 필요하다고 생각되었습니다 !


또한, 결국 코드래빗의 도움을 받아 testDB를 구현했습니다.

제가 어려움을 겪은 부분은 다음과 같아요!
```
    void findPieceByIdTest() {

        // given
        Piece soldier = new Soldier(new Position(1, 1), RED);
        pieceDao.addPiece(soldier);
        Piece piece = pieceDao.findPieceById(1);

        assertThat(piece.getClass()).isEqualTo(Soldier.class);
    }
```

다음과 같은 테스트는 1이라는 고정 값을 테스트합니다. 하지만 auto Increment 특성 상, 테이블 자체를 제거하지 않으면 기존의 결과가 남아 2라는 id로 생성을 합니다.

그래서 처음엔 성공하고, 이후에는 실패하는 경험을 했어요.

이렇게 된 이유는 도메인에 id라는 필드가 없기 때문이라고 생각하였고, 이를 해결하기 위해 현재로서는 id로 객체찾기, 객체로 id찾기를 섞어 해결하였습니다. 

>FakeDao 를 사용하여 테스트한다는것은 어떤 의미인지 이해못했어요 🤔
PieceDao 를 테스트할 건데 FakePieceDao 를 구현해서 테스트한다는것일까요?

맞아요. 근데 이부분은 저의 추측이라 현재로서는 넘어가도 될 것 같아요.

Gomding replied:
> 다음과 같은 테스트는 1이라는 고정 값을 테스트합니다. 하지만 auto Increment 특성 상, 테이블 자체를 제거하지 않으면 기존의 결과가 남아 2라는 id로 생성을 합니다.

addPiece 메서드에서 저장한 레코드의 id 를 반환받는 방법도 있을 것 같네요 :)

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -18,39 +28,88 @@ public class JanggiController {
+        return possibleRoutes;
+    }
 
+    private void movePieceIfValid(final Game game, final Piece selectedPiece, final Set<Route> possibleRoutes) {
```
> 피스를 옮길 때마다 저장? 한 번에 저장?
> 게임을 진행하면서 해당 사항을 저장하면서 가야할까요, 아니면 3번 기능(게임 종료)을 이용했을 때 일괄적으로 저장해야할까요 ?

체체는 두 방법이 어떤 장단점이 있다고 생각하시나요?

CheChe903 replied:
저장하면서 간다.

장점: 예기치 못한 오류로 게임이 종료되어도 데이터가 남아있다.
단점: 쿼리의 갯수가 증가하여 속도가 느려질 수 있다.

마지막에 저장한다.

장점: 로컬환경에서 먼저 모두 진행되므로 속도가 빠르고 데이터베이스 접근이 없어 비용이 싸다.
단점: 예기치 못한 오류로 종료된다면 저장되지 않아 사라지게 된다.

입니다.
사용자의 편의를 생각한다면 전자가 맞는 것 같아요!

Gomding commented on src/main/java/janggi/dao/PieceFactory.java:
```diff
@@ -0,0 +1,61 @@
+public class PieceFactory {
+
+    public static PieceType getPieceName(Piece piece) {
+        if (piece.getClass() == Soldier.class) {
+            return SOLDIER;
+        }
```
> 추상 클래스의 한계
> 데이터베이스에 저장하기 위해서는 해당 기물의 종류를 알아야했습니다. 그렇다면 추상 클래스를 사용하면서 데이터베이스에 저장하기 위해서는 getClass로 변환해주는 작업이 필수인 걸까요 ?

분명 어디선가 변환은 해야할것같네요!
각 기물 객체가 type 을 반환해주면 훨씬 편한 작업이 될 것 같은데 어떻게 생각하시나요?

CheChe903 replied:
그렇게 해보았습니다 !
[b4a9637](https://github.com/woowacourse/java-janggi/pull/100/commits/b4a963794bf2848c01842eb8f4e44d72c4c52504)

Gomding commented on src/main/java/janggi/dao/PieceFactory.java:
```diff
@@ -0,0 +1,61 @@
+            case ELEPHANT -> new Elephant(position, team);
+            case CHARIOT -> new Chariot(position, team);
+            case CANNON -> new Cannon(position, team);
+            default -> throw new IllegalArgumentException("알 수 없는 기물: " + pieceType);
```
꼼꼼한 예외처리 👍 👍 👍

Gomding commented on src/test/java/janggi/dao/PositionDaoTest.java:
```diff
@@ -0,0 +1,58 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class PositionDaoTest {
```
테스트를 반복 실행하는게 불가능해보이네요 😢 
각 테스트가 독립적으로 실행될 수 있게 해주시면 좋겠어요!
다른 테스트에 영향받지 않게요 :)

CheChe903 replied:
전체적으로 수정하며 되었습니다!

Gomding commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+version: "3.9"
```
docker-compose 를 활용해주셨네요 👍 👍 👍 

아래와 같이 `docker-compose.yaml` 과 같은 위치에
`./db/mysql/init/init.sql` 파일에 ddl 을 넣어두면 `docker-compose.yaml` 실행 시 바로 ddl 을 실행할 수 있어요 :)
(data 와 config 디렉토리는 무시해주세요 ㅎㅎ)

![image](https://github.com/user-attachments/assets/049ee1bf-0d71-4901-8456-66885cf9f61e)

```
-- init.sql 파일 내부

CREATE TABLE Team (
                      team_id INT PRIMARY KEY,
                      team_name VARCHAR(10) NOT NULL
);

CREATE TABLE Position (
                          position_id INT AUTO_INCREMENT PRIMARY KEY,
                          x INT NOT NULL,
                          y INT NOT NULL,
                          UNIQUE (x, y)
);


CREATE TABLE Piece (
                       piece_id INT AUTO_INCREMENT PRIMARY KEY,
                       piece_type VARCHAR(50) NOT NULL,
                       team_id INT,
                       position_id INT,
                       FOREIGN KEY (position_id) REFERENCES Position (position_id),
                       FOREIGN KEY (team_id) REFERENCES Team (team_id)
);

CREATE TABLE Turn (
    team_id int PRIMARY KEY
);

INSERT INTO Team (team_id, team_name)
VALUES
    (1, 'RED'),
    (2, 'BLUE');

```

CheChe903 replied:
감사합니다 ㅎ ㅎ

Gomding replied:
변경하면서 `docker-compose.yaml` 파일이 삭제된것 같네요 😅

Gomding commented on src/main/java/janggi/domain/piece/Piece.java:
```diff
@@ -17,38 +17,27 @@ public Piece(final Position position, final Team team) {
-                .toList();
-        return checkPiecesInRoute(route, piecesInRoute);
-    }
+    public abstract double getScore();
```
모든 기물은 점수를 가져야하니 getScore 를 추상메서드로 선언해주셨네요 💯 💯 💯

Gomding reviewed (changes requested):
안녕하세요 체체!
2단계도 잘 구현해주셨네요 💯 💯 
몇가지 코멘트 남겼으니 의견 부탁드려요!
궁금한 점 있으면 편하게 DM이나 코멘트 남겨주세요!

CheChe903 committed:
refactor(PieceType): 피스타입 기물이 반환

CheChe903 commented on src/test/java/janggi/dao/TeamDaoTest.java:
```diff
@@ -0,0 +1,22 @@
+
+    private final TeamDao teamDao = new TeamDao();
+
+    @Test
+    void findTeamById() {
+
+        // given
+        int teamId = 2;
+
+        // when
+        assertThat(teamDao.findTeamById(teamId)).isEqualTo(BLUE);
+    }
```
현재 TeamDao를 테스트하는 비즈니스 로직은 다음과 같아
```

    public Team findTeamById(int teamId) {

        final var query = "SELECT * FROM team WHERE team_id = ?";
        try (final var connection = databaseConnection.getConnection();
             final var preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setInt(1, teamId);

            final var resultSet = preparedStatement.executeQuery();
            if (resultSet.next()) {
                return Team.valueOf(resultSet.getString("team_name"));
            }
        } catch (SQLException e) {
            throw new RuntimeException();
        }
        return null;
    }
    ```
    
 여기서 return null로 처리하는 게 맞을까 ? 지금 999라는 id는 없어서 null이 반환되는 것 같아

CheChe903 replied:
고마워

CheChe903 committed:
refactor(TeamDao): 테스트 추가 및 예외 처리

- [x] 모든 팀 경우 단위 테스트를 추가합니다.
- [x] 팀이 발견되지 않을 경우 예외를 발생합니다.

CheChe903 commented on src/test/java/janggi/dao/DatabaseConnectionTest.java:
```diff
@@ -0,0 +1,25 @@
+
+    private final DatabaseConnection databaseConnection = DatabaseConnection.getInstance();
+
+    @DisplayName("커넥션 테스트")
+    @Test
+    public void connectionTest() throws SQLException {
+
+        // given
+
+        // when & then
+        try (final var connection = databaseConnection.getConnection()) {
+            assertThat(connection).isNotNull();
+        }
```
isValid 함수는 어떤 것을 하는 메서드야 ?

CheChe903 committed:
refactor(DatabaseTest): 커넥터 닫기 테스트 추가 및 var 변수 삭제

CheChe903 committed:
test(GameTest): 기물 이동 경로 테스트 검증 추가

CheChe903 committed:
test(GameTest): 기물 선택 예외 테스트 추가

CheChe903 commented on src/test/java/janggi/dao/PieceDaoTest.java:
```diff
@@ -0,0 +1,62 @@
+        }).doesNotThrowAnyException();
+    }
+
+    @DisplayName("데이터베이스에서 아이디로 기물을 찾는다.")
+    @Test
+    void findPieceByIdTest() {
+
+        // given
+        Piece soldier = new Soldier(new Position(1, 1), RED);
+        pieceDao.addPiece(soldier);
+        Piece piece = pieceDao.findPieceById(1);
+
+        assertThat(piece.getClass()).isEqualTo(Soldier.class);
+    }
```
이것을 어떻게 해결해야할 지 모르겠어 예시로,
```
    @DisplayName("데이터베이스에 기물 모음을 저장한다.")
    @Test
    void addPiecesTest() {

        // given
        Piece cannon = new Cannon(new Position(1, 1), RED);
        Piece chariot = new Chariot(new Position(2, 2), RED);
        List<Piece> pieces = List.of(cannon, chariot);

        // when & then
        assertThatCode(() -> {
            pieceDao.addPieces(pieces);
        }).doesNotThrowAnyException();
    }
```
    
이런 테스트가 있으면, Position 테이블은 x,y를 기준으로 유니크 키라 1,1에 이미 있으면 저장이 안돼 어떻게 실제 DB와 독립적으로 테스트할 수 있을까?

CheChe903 replied:
1번 방법으로 한다면 기존에 저장돼있는 실제 DB도 삭제되어서 1번으로는 해결하지 못해. 
테스트 전용 DB설정하는 법을 알려줄 수 있어?

CheChe903 replied:
@coderabbitai

CheChe903 replied:
```
    @DisplayName("데이터베이스에서 아이디로 기물을 찾는다.")
    @Test
    void findPieceByIdTest() {

        // given
        final Piece soldier = new Soldier(new Position(1, 1), RED);
        pieceDao.addPiece(soldier);
        final Piece piece = pieceDao.findPieceById(1);

        assertThat(piece.getPieceType()).isEqualTo(PieceType.SOLDIER);
    }
```

다음과 같은 테스트는 pieceId가 필요하게 돼 근데 auto_increment 때문에 새로 생성하여도 id가 1부터 시작하지 않아. 이러한 경우는 어떻게 해결할까?

CheChe903 replied:
@coderabbitai

CheChe903 replied:
그렇다면 모든 add메서드에 id를 반환하는 건 좋은 선택일까 ?

CheChe903 replied:
@coderabbitai
그렇다면 모든 add메서드에 id를 반환하는 건 좋은 선택일까 ?

CheChe903 replied:
@coderabbitai

piece를 데이터베이스에 저장할 때 positionId가 필요한데, 
그러면 Position을 저장할 때 반환을 Position 객체로 하면 Position으로 id를 찾는 로직이 필요한 거지 ?

CheChe903 replied:
@coderabbitai

현재로서는 domain인 piece나 position에 id 변수를  추가하고싶지 않아

CheChe903 committed:
test(PositionDao): 포지션 데이터베이스 테스트 수정 및 테스트 데이터베이스 생성

CheChe903 committed:
test(TurnDao): 턴 데이터베이스 테스트 구현

CheChe903 committed:
feat(PositionDao): 오류 메시지 수정

CheChe903 committed:
test(PieceDaoTest): 기물 데이터베이스 로직 테스트 추가

CheChe903 committed:
docs(README): 리드미 업데이트

CheChe903 committed:
fix(PositionDao): 객체 존재하지 않는 버그 수정

- [x] Position이 데이터베이스에 존재하지 않으면 Optional을 반환하도록 합니다.

CheChe903 committed:
fix(PositionDao): 객체 존재하지 않는 버그 수정

- [x] Position이 데이터베이스에 존재하지 않으면 Optional을 반환하도록 합니다.

CheChe903 committed:
fix: 메서드 시그니처 변경

CheChe903 committed:
fix(Controller): 게임 포기 메서드 네이밍 변경

CheChe903 committed:
fix(Controller): 불필요한 DB 접근 삭제

CheChe903 committed:
docs(docker-compose): db로 이동

CheChe903 commented:
@coderabbitai full review

Gomding commented on src/main/java/janggi/dao/DatabaseConnection.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    public void closeConnection(final Connection connection) {
```
`closeConnection` 은 사용하는곳이 테스트밖에 없는데 구현하신 이유가 궁금해요!

CheChe903 replied:
코드래빗이 추천하여 진행해보았는데 생각해보니 그렇네요!
너무 코드래빗을 따라가는 것도 좋지 않은 것 같아요.

[a533f55](https://github.com/woowacourse/java-janggi/pull/100/commits/a533f55176d2c5bafa6b2d7574d2498dd64def88)

Gomding commented on src/main/java/janggi/dao/DatabaseConnection.java:
```diff
@@ -0,0 +1,54 @@
+            final String url = "jdbc:mysql://" + SERVER + "/" + currentDb + OPTION;
+            return DriverManager.getConnection(url, USERNAME, PASSWORD);
+        } catch (final SQLException e) {
+            throw new RuntimeException("오류가 발생했습니다.");
```
예외 메세지는 더 친절하게 작성해주시는게 좋아요!

- https://hyeon9mak.github.io/why-should-throw-detail-error-message/
- https://hyeon9mak.github.io/writing-helpful-error-messages/

CheChe903 replied:
현재 수정한 시점에서 다시 한 번 리뷰 부탁드려도 될까요 ?!

Gomding replied:
물론이죠 😄 

> throw new RuntimeException("데이터 베이스 연결에 실패했습니다.", e);

잘 수정해주신것 확인했어요!!

Gomding commented on src/main/java/janggi/dao/TeamDao.java:
```diff
@@ -0,0 +1,45 @@
+        }
+    }
+
+    public int findIdByTeam(final Team currentTeam) {
```
인자 이름이 `currentTeam` 일 이유가 있을까요?

Gomding commented on src/test/java/janggi/dao/PieceDaoTest.java:
```diff
@@ -0,0 +1,98 @@
+
+        // when
+        final int findPieceId = pieceDao.findIdByPiece(solider);
+        final Piece findPiece = pieceDao.findPieceById(findPieceId);
```
`pieceDao.findPieceById(findPieceId);` 해당 조회가 필요한 이유는 무엇일까요? 🤔 
해당 테스트의 목적은 `기물의 id를 찾는다.` 가 아니었을까요?

CheChe903 replied:
맞습니다! 다만, id를 반환하지 않는 그 상태에서는 저게 최선이었다고 생각합니다..
이젠 아이디를 반환하게 됩니다!

Gomding commented on src/test/java/janggi/dao/PieceDaoTest.java:
```diff
@@ -0,0 +1,98 @@
+        final Piece piece = new Soldier(new Position(1, 1), RED);
+
+        // when & then
+        assertThatCode(() -> {
+            pieceDao.addPiece(piece);
+        }).doesNotThrowAnyException();
```
예외가 발생하지 않았으면 정상 저장됐다고 판단해도 괜찮을까요! 😃

CheChe903 replied:
위와 동일합니다!

Gomding commented on src/main/java/janggi/dao/DatabaseConnection.java:
```diff
@@ -0,0 +1,54 @@
+
+    public Connection getConnection() {
+        try {
+            final String currentDb = isTestMode ? TEST_DATABASE : DATABASE;
```
프로덕션 코드에 테스트용 코드가 들어갔군요 🤔 
체체는 테스트 용 설정이 들어가있는것을 어떻게 생각하시나요?
테스트 환경을 변경하고싶으면 프로덕션 코드를 수정해야하는군요 😢 

특히 테스트에서 setTestMode 를 잘못 설정하면 프로덕션 데이터베이스에 데이터가 들어갈 수 있겠군요 🤔

CheChe903 replied:
수정하며 삭제되었습니다 :)

Gomding commented on src/test/java/janggi/dao/PositionDaoTest.java:
```diff
@@ -0,0 +1,78 @@
+    void setUp() {
+        pieceDao.deleteAllPieces();
+        positionDao.deleteAllPositions();
+    }
```
각 테스트 실행 전에 데이터를 전부 삭제하는 방식으로 테스트간 독립을 구현해주셨네요 👍 👍 👍

Gomding commented on src/test/java/janggi/dao/PositionDaoTest.java:
```diff
@@ -0,0 +1,78 @@
+        positionDao.deleteAllPositions();
+    }
+
+    @AfterAll
+    static void tearDownClass() {
+        DatabaseConnection.setTestMode(false);
+    }
```
해당 부분이 필요한 이유는 무엇일까요?

CheChe903 replied:
TestMode를 꺼주기 위함이었습니다! 수정되며 삭제되었습니다. 실제와 테스트가 분리되지 않아 생긴 메서드였습니다.

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -18,39 +28,91 @@ public class JanggiController {
+
+    }
+
+    private boolean checkGameOver(final Game game) {
+        return game.getStatus() == PROGRESS;
     }
```
메서드 이름이 애매한 것 같아요!

- 이름 : 게임이 끝났는지 확인
- 행위 : 진행중인가?

+) check는 단순 검사 메서드(내부에서 검사하여 적절하지않으면 예외 던짐)로 보이기도해서
is~~ 로 작성해주시는게 좋다고 생각해요!

CheChe903 replied:
수정되었습니다! 

1c95dd6

Gomding commented on README.md:
```diff
@@ -47,89 +54,63 @@
+- [x] 서로가 게임 종료에 동의하면 남아있는 기물의 점수를 계산 후 출력한다.
+    - [x] 해당 턴이 아닌 사람은 1.5의 가산 점수를 받는다.
+
+## 데이터베이스 사용법
```
사용법도 명시해주신 점 너무 친절하네요 👍 👍 👍

Gomding commented on src/main/java/janggi/domain/Function.java:
```diff
@@ -0,0 +1,7 @@
+package janggi.domain;
+
+public enum Function {
```
해당 객체는 도메인에서 사용하는것이 맞을까요?

CheChe903 replied:
인풋뷰와 컨트롤러에서만 사용되는데, 혹시 패키지 때문일까요 ? ?

Gomding replied:
패키지 때문에 맞습니다 :)
domain 에서 사용하지 않는 요소인데 domain 에 있으면 혼란을 줄 것 같아서요

+) 추가로 `Function` 이라는 이름은 java 의 함수형 인터페이스 `Function` 과 헷갈릴 수 있을것 같아요 😢 
해당 이름을 꼭 사용해야하는게 아니면 다른 이름을 고려해볼 수도 있어요!

CheChe903 replied:
현재 JanggiFunction으로 바꾸었고, 해당 객체는 컨트롤러와 뷰에서만 사용되기에 컨트롤러로 옮겼습니다!
만약 다른 컨트롤러가 생긴다면 다른 패키지를 또 고려보해볼 것 같아요.


[617bd9c](https://github.com/woowacourse/java-janggi/pull/100/commits/617bd9ce29c109ea275318c46efb54062a31c252)

Gomding commented on src/main/java/janggi/dao/PieceDao.java:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    public int findIdByPiece(final Piece piece) {
+        final String query = "SELECT * FROM piece WHERE team_id = ? AND position_id = ? AND piece_type = ?";
```
필요한건 team_id, position_id, piece_type 3가지인것 같은데
Piece 전체를 인자로 받은 이유가 있을까요?

CheChe903 replied:
이부분에 궁금한 것이 있어요.

Controller -> Service -> Repository -> Dao 순으로 계층이 나뉜다고하면,  
Controller -> Service  -> Repository 도메인 사용
Repository -> Dao 에서는 원시값 또는 엔티티를 사용하는 것이 맞는 지 알고 싶어요.

Gomding replied:
다른 코멘트에서도 언급했지만 Repository 와 DAO 의 차이는 무엇일까요? 😃 

+) 원시값과 엔티티는 어떤 차이가 있나요?

++) 체체가 말하는 엔티티란 무엇일까요?

CheChe903 replied:
> Repository 와 DAO 의 차이는 무엇일까요?

 DAO는 데이터베이스와의 연결을 담당하는 데이터 액세스에 초점을 맞추고, Repository는 도메인 모델을 중심으로 데이터를 관리하는 추상 계층이라고 생각해요. 
 
현재로서의 제 의견은, Dao는 원시값 또는 엔티티를 사용해도 무방하다고 생각합니다.

> 원시값과 엔티티는 어떤 차이가 있나요?

원시값은 말 그대로 원시값, (int, long등 )이고, 엔티티는 데이터베이스에 저장할 수 있는 상태라고 생각합니다. 좀 더 개념적으로 접근하자면 엔티티는 우리의 프로젝트가 저장할 수 있는 규칙이라고도 생각이 들어요.

> 체체가 말하는 엔티티란 무엇일까요?

위에서 말씀드린 것처럼 해당 프로젝트가 나타내는 테이블을 보여주기 위함이며 저장할 수 있는 구조라고 생각합니다!

Gomding replied:
> DAO는 데이터베이스와의 연결을 담당하는 데이터 액세스에 초점을 맞추고, Repository는 도메인 모델을 중심으로 데이터를 관리하는 추상 계층이라고 생각해요.

`DAO는 데이터베이스와의 연결을 담당하는 데이터 액세스에 초점을 맞추고` 👍 👍 👍 
도메인 모델을 중심으로 데이터를 관리하면 어떤 장점이 있을까요?

> 위에서 말씀드린 것처럼 해당 프로젝트가 나타내는 테이블을 보여주기 위함이며 저장할 수 있는 구조라고 생각합니다!

테이블 구조를 보여주기 위함이군요
체체가 말하는 Entity가 DTO와는 어떤 차이점을 가지고 있을까요?
구분해야하는 개념인지 궁금해요!

CheChe903 replied:
먼저, 같은 점을 따지자면 엔티티는 서버 <-> Database를 매핑하는 DTO역할이라고 생각합니다.
하지만 엄연히 엔티티는 고유적으로 식별할 수 있는 식별자를 갖습니다. (id같은 경우)
DTO는 값들만 담겨 있으므로 값들이 모두 동일하면 동일하다고 판단하게 됩니다.

Gomding commented on src/main/java/janggi/dao/PieceDao.java:
```diff
@@ -0,0 +1,143 @@
+        final String query = "SELECT * FROM piece WHERE team_id = ? AND position_id = ? AND piece_type = ?";
+        try (final Connection connection = databaseConnection.getConnection();
+             final PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            final int teamId = teamDao.findIdByTeam(piece.getTeam());
+            preparedStatement.setInt(1, teamId);
+            final int positionId = positionDao.findIdByPosition(piece.getPosition()).get();
```
PieceDao는 Piece 에 대한 DB 접근을 하는 책임을 가졌다고 생각해요
다른 DAO를 의존하고 호출해도 되는걸까요?

Gomding reviewed (changes requested):
안녕하세요 체체!
피드백 잘 반영해주셨네요 👍 
추가로 코멘트 남겼으니 확인부탁드려요!!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요~

CheChe903 committed:
feat(Database): 데이터베이스 관련 로직 삭제

CheChe903 committed:
feat(QueryProcessor): 데이터베이스 로직 실행 구현

CheChe903 committed:
docs(docker): 도커 및 init 파일 수정

CheChe903 committed:
docs(README): 데이터베이스 사용법 업데이트

CheChe903 committed:
feat: Database 로직 구현

CheChe903 committed:
feat(DatabaseConnector): 커넥터 인터페이스 구현

CheChe903 committed:
feat(Fake): 페이크 객체 구현 및 테스트 추가

CheChe903 committed:
test(JdbcRepositoryTest): 데이터베이스 테스트 구현

CheChe903 committed:
fix(DatabaseConnection): close 삭제

CheChe903 committed:
fix(DatabaseConnection): 에러 메시지 수정

CheChe903 committed:
refactor(Controller): checkGameOver 메서드 변경

- isGameOver로 변경합니다.

CheChe903 commented:
@coderabbit ai full review

Gomding commented on src/main/java/janggi/service/PieceService.java:
```diff
@@ -0,0 +1,35 @@
+import janggi.repository.PieceRepository;
+import java.util.List;
+
+public class PieceService {
```
체체가 생각하는 `PieceService`, `TurnService` 객체의 책임은 무엇인가요?
제가 볼때는 단순 Repository 메서드 하나만 호출하는 바이패스 메서드만 존재하는것으로 보여서요!

CheChe903 replied:
서비스 계층의 의미를 다시 생각해보니 컨트롤러와 레파지토리를 잇는 역할이라고 생각합니다.
그렇게 맞춰 컨트롤러를 수정하고, 두 서비스를 합친 `GameService`를 만들어 보았어요.

CheChe903 replied:
[764b060](https://github.com/woowacourse/java-janggi/pull/100/commits/764b060f1a340643a3f90c0782b558a59be5b399)

Gomding commented on src/main/java/janggi/repository/JdbcPieceRepository.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class JdbcPieceRepository implements PieceRepository {
+
+    private final PieceDao pieceDao;
```
Dao 와 Repository 는 어떤 차이가 있나요?

Gomding commented on src/main/java/janggi/repository/JdbcPieceRepository.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class JdbcPieceRepository implements PieceRepository {
```
해당 객체안에 JDBC 개념이 들어있나요? 🤔

CheChe903 replied:
어떻게 바꾸면 좋을까요 ? DefaultPieceRepository로 해도 충분할까요 ?

Gomding replied:
#체체가 옳다고 생각하는 이름으로 변경해주시면 좋을것 같아요 :)

의견 드린이유는 Jdbc와는 내부를 타고들어가야 의존이 있는데 
상위 객체에 해당 개념이 있는것은 어색했어요 :)

+) 해당 클래스의 이름이 애매한 이유는 DAO 를 한 번 감싸줬기 때문이라 생각하는데
JdbcPieceRepository 가 `QueryProcessor` 를 바로 의존하는것은 어떻게 생각하시나요?

CheChe903 replied:
>해당 클래스의 이름이 애매한 이유는 DAO 를 한 번 감싸줬기 때문이라 생각하는데
JdbcPieceRepository 가 QueryProcessor 를 바로 의존하는것은 어떻게 생각하시나요?

중간에 PieceDao가 있기 때문에 Jdbc라는 네이밍이 어울리지 않는 거군요, 
만약 JdbcPieceRepository -> QueryProcessor -> DB 연결 이라면 해당 네이밍이 이해가 되겠네요!
현재 구조로서는 연결하는 것은 dao인데, jdbc가 레포지토리 네이밍이 있으니 다른 사람이 보았을 땐 혼란을 줄 수 있을 것 같아요.

CheChe903 replied:
[6ff73ff](https://github.com/woowacourse/java-janggi/pull/100/commits/6ff73fffe01ad2d18ef05f8b3fae5d30c88ab859)

Gomding commented on src/test/java/janggi/repository/FakePieceRepository.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.ArrayList;
+import java.util.List;
+
+class FakePieceRepository implements PieceRepository {
```
FakePieceRepository 는 무엇을 위해 생성하신건지 궁금해요! 🙋

CheChe903 replied:
Fake는 MySQL이 연결되어 있지 않아도 레포지토리에 대한 테스트를 독립적으로 받아볼 수 있는 환경입니다!
또한 데이터베이스를 연결하는 테스트는 속도가 느리더라구요,
테스트의 이점은 빠른 피드백을 받는 것인데 MySQL이 연결되지 않았다면 빠른 피드백을 받을 수 없기 때문에 작성해보았습니다.

Gomding commented on src/test/java/janggi/repository/FakePieceRepositoryTest.java:
```diff
@@ -0,0 +1,99 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class FakePieceRepositoryTest {
```
Fake 객체를 테스트하는 목적도 궁금해요!

CheChe903 replied:
위와 동일하게, 테스트의 이점은 빠른 피드백을 받는 것인데 MySQL이 연결되지 않았다면 빠른 피드백을 받을 수 없기 때문에 작성해보았습니다!!

Gomding replied:
> 위와 동일하게, 테스트의 이점은 빠른 피드백을 받는 것인데 MySQL이 연결되지 않았다면 빠른 피드백을 받을 수 없기 때문에 작성해보았습니다!!

말씀하신 것이 Fake 객체를 테스트하는 이유가 맞을까요? 😳

질문드린 의도는 `테스트를 위해 만든 객체를 테스트한다` 라는게 어색하게 느껴져서요 🤔

CheChe903 replied:
이 부분에 대해서는 이해가 잘 가지 않은 것 같아요.
저는 `테스트를 위해 만든 객체를 테스트한다.`가 어색하게 느껴지지 않는데, 어떠한 부분에서 어색하게 느껴지는 지 알 수 있을까요 ?

결과적으로 DB연결 의존성 없이 테스트 하기 위해 만든 Fake객체를 테스트하지 않으면 Fake 객체를 만들 이유도 없지 않을까? 라는 생각이 들어요.

Gomding replied:
테스트 코드는 보통 프로덕션 코드에 대한 검증인데
체체가 만들어주신 Fake 객체는 테스트에서 외부 영향을 끊어내기 위해 구현했다고 생각해요
체체가 프로덕션 코드에서 사용하지않고 조작하여 만든 Fake 객체를 테스트하는것은 어떤 의미가 있는지 궁금했어요!

CheChe903 replied:
프로덕션에서 사용하지 않지만, 데이터베이스의 의존성이 있는 레파지토리를 인메모리를 이용하여 새롭게 정의하여 그 레파지토리는 무슨 일을 한다, 할 수 있다는 것을 문서처럼 보여주기 위해 작성하였다고 생각합니다!

Gomding commented on src/test/java/janggi/repository/JdbcPieceRepositoryTest.java:
```diff
@@ -0,0 +1,104 @@
+
+class JdbcPieceRepositoryTest {
+    
+    private final PieceRepository pieceRepository = new JdbcPieceRepository(
+            new PieceDao(new QueryProcessor(TestMySQLDatabaseConnection.getInstance())));
```
> new QueryProcessor(TestMySQLDatabaseConnection.getInstance()))

테스트용 커넥션을 넣은 QueryProcessor 객체를 저장소 테스트마다 매번 생성해야하니 
`Test Fixture` 같은것을 사용해보면 좋겠어요 :)

CheChe903 replied:
해보았습니다 :)

[d3cc4f2](https://github.com/woowacourse/java-janggi/pull/100/commits/d3cc4f270158bc7ef09b7026aee66d098d4a4437)

Gomding replied:
코멘트마다 반영 커밋을 남겨주시는 습관이 정말 좋다고 생각해요 👍 👍 👍 
귀찮고 사소하지만 리뷰어에게 너무나 친절한 행위입니다 ㅎㅎ

CheChe903 replied:
감사합니다 !!

Gomding commented on src/main/java/janggi/database/dao/PieceDao.java:
```diff
@@ -0,0 +1,39 @@
+import janggi.database.entity.PieceEntity;
+import java.util.List;
+
+public class PieceDao {
```
Dao 테스트는 진행하지 않으신 이유가 궁금해요!
테스트 작성 기준을 논의해보면 좋겠군요 🤗

CheChe903 replied:
앗 실수로 누락되었네요 ㅎㅎ.. 추가하였습니다!

CheChe903 replied:
[afb83d8](https://github.com/woowacourse/java-janggi/pull/100/commits/afb83d83a6e05d40638de389866f4612d9827409)

Gomding commented on src/main/java/janggi/database/MySQLDatabaseConnection.java:
```diff
@@ -0,0 +1,37 @@
+    @Override
+    public Connection getConnection() {
+        try {
+            final String url = "jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION;
```
URL 상수가 있는데 문자열을 다시 생성하신 이유가 궁금해요!

CheChe903 replied:
이것도 실수입니다!! ..

CheChe903 replied:
[4eaf5c9](https://github.com/woowacourse/java-janggi/pull/100/commits/4eaf5c979163bc305b974f2005f32ba250f4170a)

Gomding commented on src/main/java/janggi/database/QueryProcessor.java:
```diff
@@ -0,0 +1,111 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryProcessor {
```
`QueryProcessor` 구현 훌륭하네요 👍 👍 👍

Gomding commented on src/main/java/janggi/database/QueryProcessor.java:
```diff
@@ -0,0 +1,111 @@
+            }
+            throw new SQLException("키 생성에 실패했습니다.");
+        } catch (final SQLException e) {
+            throw new RuntimeException("데이터베이스 작업 중 오류 발생: " + query, e);
```
QueryProcessor 에서만 중복해서 발생하는 예외 메세지가 여럿있네요!
커스텀 예외 클래스를 생성해서 클래스이름으로 예외상황을 표현하고 메세지 중복을 한 곳에서 관리할 수 있겠어요 :)

CheChe903 replied:
예전 프리코스 때 커스텀 예외를 만들어봤는데, 필요성을 못찾아서 이후에는 하지 않았어요, 
현재로서는 데이터베이스 예외는 따로 처리해주는 것이 가독성이 더 좋을 것 같아요(외부 요소라고 생각되어서 입니다!).

CheChe903 replied:
[98958c8](https://github.com/woowacourse/java-janggi/pull/100/commits/98958c8b62887a440b1ee1af18073bea382bb990)

Gomding replied:
> 현재로서는 데이터베이스 예외는 따로 처리해주는 것이 가독성이 더 좋을 것 같아요(외부 요소라고 생각되어서 입니다!).

해당 기준도 좋네요! 👍 👍

Gomding reviewed (changes requested):
안녕하세요 체체!
고민하신 부분을 잘 반영해주셨네요 👍 👍 
추가로 코멘트 남겼으니 확인부탁드려요~

궁금한점 있으면 언제든 DM이나 코멘트 남겨주세요 😄

CheChe903 committed:
refactor(Function): Function 이넘 클래스 컨트롤러로 이동

- [x] 컨트롤러와 뷰에서 사용되기 때문에 controller 패키지에 둡니다.

CheChe903 committed:
fix(FakeTurnRepository): id 필드 접근제어자 변경

CheChe903 committed:
docs(init.sql): Team 테이블 구조 변경

CheChe903 committed:
refactor(Controller): isGameOver 메서드명 변경

CheChe903 committed:
refactor(Controller, GameService): 컨트롤러 책임 서비스로 이전

CheChe903 committed:
feat(TestFixture): 데이터베이스 연동 테스트를 위한 픽스쳐 추가

CheChe903 committed:
test(Dao): Dao 테스트 추가

CheChe903 committed:
refactor(DatabaseConnection): url 상수 변환

CheChe903 committed:
feat(DatabaseException): 데이터베이스 커스텀 예외 작성

Gomding commented on src/main/java/janggi/controller/JanggiFunction.java:
```diff
@@ -0,0 +1,7 @@
+package janggi.controller;
+
+public enum JanggiFunction {
```
> 만약 다른 컨트롤러가 생긴다면 다른 패키지를 또 고려보해볼 것 같아요.

추가적인 기준도 좋다고 생각해요! 💯 💯 💯

Gomding commented on src/main/java/janggi/service/PieceService.java:
```diff
@@ -0,0 +1,35 @@
+import janggi.repository.PieceRepository;
+import java.util.List;
+
+public class PieceService {
```
사용하지 않는 서비스 클래스는 제거해도 되겠군요! 😃

CheChe903 replied:
수정입니다 !
[df2df81](https://github.com/woowacourse/java-janggi/pull/100/commits/df2df815c0f76d78ea309aedafc7da225a37ba33)

Gomding commented on src/main/java/janggi/service/GameService.java:
```diff
@@ -0,0 +1,71 @@
+import janggi.repository.TurnRepository;
+import java.util.List;
+
+public class GameService {
```
두 서비스를 합쳐주신것은 좋은 변경이라 생각해요 👍 👍 
체체가 생각하는 서비스의 역할도 공유해주시면 좋겠어요 :)

CheChe903 replied:
현재로서 서비스의 역할은 데이터베이스와의 연결이 가장 중요 포인트라고 생각합니다! 
결과적으로          

```
뷰 ---------- 컨트롤러 ----- 도메인
                 ㅣ------- 서비스 -------DB
```

가 된다고 생각합니다. 게임을 불러오고, 저장하고, 확인하는 역할입니다.

Gomding commented on src/main/java/janggi/service/GameService.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public Game loadGame() {
+        if (turnRepository.find().isPresent()) {
+            return new Game(new Pieces(pieceRepository.findAll()), turnRepository.find().get());
```
`turnRepository.find()` 를 두 번 호출하는 이유가 있을까요 :)

+) find 메서드가 바로 알기 어려울 것 같아요 😢 
만약 체체가 말하는 Repository 가 도메인 저장소를 추상화한 것이라면
이름도 Turn 도메인에 맞게 명확하게 해보시면 좋겠어요!

CheChe903 replied:
바꿔보았습니다!

[a0c351d](https://github.com/woowacourse/java-janggi/pull/100/commits/a0c351d270ae158a8f1cdf5f6888627058c8d86c)

Gomding reviewed (changes requested):
안녕하세요 체체!
빠르게 반영해주셨네요 :)
이제 거의 끝나가는것 같아요 😄 
몇 가지만 더 논의해보면 좋겠어요!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요~

CheChe903 committed:
refactor(Repository): 구현체 저장소 클래스명 변경

- Jdbc의 개념이 없기 때문에 기본 구현체를 나타내야합니다. 따라서 JdbcRepository -> DefaultRepository로 변경합니다.

CheChe903 committed:
refactor(Service): 사용하지 않는 서비스 삭제

CheChe903 committed:
refactor(TurnRepository): 메서드 시그니처 변경

Gomding commented on src/main/java/janggi/controller/JanggiController.java:
```diff
@@ -18,39 +23,88 @@ public class JanggiController {
         if (canMove(possibleRoutes, destination)) {
-            board.movePiece(destination, selectPiece);
-            board.changeTurn();
+            gameService.deletePiece(selectedPiece);
+            gameService.deletePieceByPosition(destination);
+            game.movePiece(destination, selectedPiece);
+            gameService.addPiece(selectedPiece);
+            game.changeTurn();
+            gameService.updateTurn(game.getTurn());
```
> 현재로서 서비스의 역할은 데이터베이스와의 연결이 가장 중요 포인트라고 생각합니다!

해당 답변에 대한 연장선입니다!
GameService 에 비즈니스 흐름을 담아보는것은 어떻게 생각하시나요?
`이동과 턴 넘김 + 저장소 제어` 를 하나의 흐름으로 묶어주는것이죠 :)

CheChe903 replied:
```
game.movePiece(destination, selectedPiece);
```

이 로직도 service로 옮기라는 말씀이실까요 ?

현재 제가 저렇게 코드를 작성한 이유는 
```
game.movePiece(destination, selectedPiece);
```
해당 로직이 실행되면 이미 selectedPiece는의포지션이 새로운 포지션(destination)으로 바뀌며 그 자리에 있던 기물은 List에서 삭제됩니다.
따라서 destination에 이미 옮기려던 기물이 있게 됩니다.
따라서 움직이기 전에 두 기물을 삭제한 후, 로컬 상에서 움직이고, 움직임 당한 기물을 새로 데이터베이스에 저장하는 형식으로 하였습니다!

Gomding replied:
해당 흐름은 기물을 움직이기위한 `DB 접근과 도메인 객체 호출` 을 하고있는데
이 부분은 Service 내부에서 하나의 메서드로 만들어도 된다고 생각해요!

현재는 Controller 에 해당 흐름이 있는데 만약 Controller 를 새로 구현한다면(콘솔에서 웹으로 전환 등) 해당 흐름을 다시 구현해야돼요
만약 해당 흐름이 Service 에 들어있었다면 새로운 Controller 에서도 Service 만 의존하고 있다면 동일한 흐름을 쉽게 구현할 수 있겠죠 :)

Gomding reviewed (changes requested):
안녕하세요 체체!
머지해도 되는 상태인데 추가로 궁금하신 부분이 있으면 말씀해주세요~
없다면 추가로 남긴 코멘트만 확인하시고 바로 재요청 주셔도 됩니다 😄

Gomding reviewed (approved):
안녕하세요 체체!
장기 미션 진행하느라 고생하셨어요 😃 
말씀드린대로 이만 머지하겠습니다~!
레벨1 고생하셨어요 🎉 🎉 

추가로 궁금하신 점 있으면 언제든 DM 이나 코멘트 남겨주세요
방학 잘 보내세요~ 😌
